README
NAME

    constant - Perl pragma to declare constants


DESCRIPTION

    This pragma allows you to declare constants at compile-time.

    When a constant is used in an expression, Perl replaces it with its
    value at compile time, and may then optimize the expression further.
    In particular, any code in an "if (CONSTANT)" block will be optimized
    away if the constant is false.


INSTALLATION

    To install this module, run the following commands:

        perl Makefile.PL
        make
        make test
        make install

    Alternatively, to install with Module::Build, you can use the 
    following commands:

        perl Build.PL
        ./Build
        ./Build test
        ./Build install


SUPPORT AND DOCUMENTATION

    After installing, you can find documentation for this module with 
    the perldoc command.

        perldoc constant

    You can also look for information at:

        Search CPAN
            http://search.cpan.org/dist/constant

        CPAN Request Tracker:
            http://rt.cpan.org/NoAuth/Bugs.html?Dist=constant

        AnnoCPAN, annotated CPAN documentation:
            http://annocpan.org/dist/constant

        CPAN Ratings:
            http://cpanratings.perl.org/d/constant


COPYRIGHT AND LICENCE

    Copyright (C) 1997, 1999 Tom Phoenix.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.


Changes
Revision history for constant

1.27    2013.03.21    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Updated from bleadperl:
         - Remove a reference to a non-existent module (Karl Williamson)
         - Make utf8 downgrading version dependent in order to avoid
           unconditional loading of utf8_heavy.pl (Brad Gilbert)

1.25    2013.01.19    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Updated from bleadperl:
         - Remove 5.6-isms (Sébastien Aperghis-Tramoni)

1.24    2013.01.07    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Updated from bleadperl:
         - Disable the UTF8 downgrade when unnecessary (Father Chrysostomos)
         - Restore compatibility with Perl 5.8 (Sébastien Aperghis-Tramoni)
         - Skip t/utf8.t on Perl before 5.8.4 (Sébastien Aperghis-Tramoni)
        [DIST] Perl-RT #114050: Added meta-information in Makefile.PL to
        reflect that the module is maintained by P5P.

1.21    2011.04.18    SAPER (Sébastien Aperghis-Tramoni) #PerlQA2011
        [CODE] Updated from bleadperl:
         - Make Unicode constants under use utf8 work again (Father Chrysostomos)
        [TESTS] Make tests pass under old Perls.

1.20    2011.04.16    SAPER (Sébastien Aperghis-Tramoni) #PerlQA2011
        [CODE] Updated from bleadperl:
         - [perl #68640] Wrong error for undef constant name (Zefram).
        [DIST] CPAN-RT #66727: Add explicit configure_requires to Build.PL
        (David Golden).

1.19    2009.09.12    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Optimisation: Avoid unnecessary flushing of the caller
        package's MRO cache (Nicholas Clark).
        [CODE] Optimisation: Use the internal constant _CAN_PCS to optimise
        out things that can't happen pre 5.10 (Nicholas Clark).

1.18    2009.08.18    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Optimisation: Generate perl version-dependent regexps once,
        rather than every call to import (Nicholas Clark).
        [CODE] Optimisation: Convert perl version check in import() from
        run time to compile time (Nicholas Clark).
        [TESTS] Re-added pseudohashes tests that were removed at some point
        from bleadperl (Nicholas Clark).

1.17    2008.10.29    SAPER (Sébastien Aperghis-Tramoni)
        [TESTS] Moved the TRAILING test from t/constant.t to t/more-tests.t
        in order to avoid FAIL reports when executed with Test::More 0.82+

1.16    2008.09.21    SAPER (Sébastien Aperghis-Tramoni)
        [DOC] The copyright section is also the license.
        [DOC] Added a SEE ALSO section.
        [DOC] Converted files to UTF-8.
        [DOC] Improved the description in README.

1.15    2007.12.31    SAPER (Sébastien Aperghis-Tramoni)
        [TEST] Adjusted t/constant.t to how empty prototypes are reported
        among different versions of Perl (thus fixing the FAIL reports for 
        Perl 5.8.0 to 5.8.3).

1.14    2007.12.27    SAPER (Sébastien Aperghis-Tramoni)
        [DIST] CPAN-RT#31627: Specify core install dir in Build.PL. 
        Thanks to Michael G Schwern.

1.13    2007.12.27    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Merged changed blead@32338: UNITCHECK is only a keyword 
        post 5.009 (Nicholas Clark).
        [DIST] Copied the sysnopsis into eg/

1.12    2007.11.06    SAPER (Sébastien Aperghis-Tramoni)
        [DIST] CPAN-RT#30460: The module must be installed in core, not in
        site (because it get shadowed). Thanks to Imacat.

1.11    2007.09.25    SAPER (Sébastien Aperghis-Tramoni)
        [CODE] Fixed code and tests so everything work under Perl 5.005.
        [DOC] Some Pod nits. Added information about current maintainers.

1.10    2007.04.19
        blead@30980: C3 MRO support (Brandon L Black).
        blead@30824: Small fix for Symbian (Jarkko Hietaniemi).

1.09    2007.03.13
        blead@30561: Inform constant.pm about UNITCHECK (Joshua ben Jore).

1.08    2007.02.13
        blead@30255: Remove some debugging code (Nicholas Clark).
        blead@26502: Use Internals::inc_sub_generation() to invalidate 
          cached methods (Nicholas Clark).

1.07    2005.12.26
        blead@26487: Rework constant.pm to take advantage of the space 
          savings of proxy constant subroutines whenever it can (Nicholas Clark).
        blead@26485: Factore caller() call out of the loop (Nicholas Clark).

1.06    2005.12.24
        blead@26483: Using a hash reference rather than a hash can save a 
          hash copy (Nicholas Clark).

1.05    2004.11.12
        blead@23497: Silence an unnecessary warning (Anno Siegel).

1.04    2001.07.29
        blead@12097, 11473: Major POD rewrite (Ilmari Karonen).

1.03    2001.06.29
        blead@8240: Declare multiple constants at once (Casey West).

1.02    2000.03.13
        blead@5702: Use warnings::register (Paul Marquess).
        blead@5463: Relax restrictions on constant names a bit (M J T Guy).
        blead@4905: STOP blocks have been renamed to CHECK (Gurusamy Sarathy).

1.01    1999.12.28
        blead@4718: New version of constant.pm from Tom Phoenix; added Tom's 
          notes to perldelta; added STOP, DESTROY and AUTOLOAD to specials list
        blead@2029: Restore sanity to "constant" references (Gurusamy Sarathy).


Full history:
    http://public.activestate.com/cgi-bin/perlbrowse/l/lib/constant.pm



Makefile.PL
use 5.006;
use strict;
use ExtUtils::MakeMaker;

my %prereq = (
    "Test::More" => 0,
);

$prereq{"warnings::compat"} = 0 if $] < 5.006;

WriteMakefile(
    NAME            => 'constant',
    LICENSE         => 'perl',
    AUTHOR          => 'Sebastien Aperghis-Tramoni <sebastien@aperghis.net>',
    VERSION_FROM    => 'lib/constant.pm',
    ABSTRACT_FROM   => 'lib/constant.pm',
    INSTALLDIRS     => 'perl',
    PL_FILES        => {},
    PREREQ_PM       => \%prereq,
    META_MERGE          => {
        resources       => {
            repository  => "git://perl5.git.perl.org/perl.git",
            license     => 'http://dev.perl.org/licenses/',
            homepage    => 'https://metacpan.org/module/constant',
            irc         => 'irc://irc.perl.org/#p5p',
            mailinglist => 'http://lists.perl.org/list/perl5-porters.html',
            bugtracker  => "https://rt.perl.org/rt3/Search/Results.html?Query=Queue='perl5' AND Content LIKE 'module=constant' AND (Status='open' OR Status='new' OR Status='stalled')",
        },
    },
    dist            => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean           => { FILES => 'constant-*' },
);


lib/constant.pm
package constant;
use 5.005;
use strict;
use warnings::register;

use vars qw($VERSION %declared);
$VERSION = '1.14';

#=======================================================================

# Some names are evil choices.
my %keywords = map +($_, 1), qw{ BEGIN INIT CHECK END DESTROY AUTOLOAD };
$keywords{UNITCHECK}++ if $] > 5.009;

my %forced_into_main = map +($_, 1),
    qw{ STDIN STDOUT STDERR ARGV ARGVOUT ENV INC SIG };

my %forbidden = (%keywords, %forced_into_main);

#=======================================================================
# import() - import symbols into user's namespace
#
# What we actually do is define a function in the caller's namespace
# which returns the value. The function we create will normally
# be inlined as a constant, thereby avoiding further sub calling 
# overhead.
#=======================================================================
sub import {
    my $class = shift;
    return unless @_;			# Ignore 'use constant;'
    my $constants;
    my $multiple  = ref $_[0];
    my $pkg = caller;
    my $symtab;
    my $str_end = $] >= 5.006 ? "\\z" : "\\Z";

    if ($] > 5.009002) {
	no strict 'refs';
	$symtab = \%{$pkg . '::'};
    };

    if ( $multiple ) {
	if (ref $_[0] ne 'HASH') {
	    require Carp;
	    Carp::croak("Invalid reference type '".ref(shift)."' not 'HASH'");
	}
	$constants = shift;
    } else {
	$constants->{+shift} = undef;
    }

    foreach my $name ( keys %$constants ) {
	unless (defined $name) {
	    require Carp;
	    Carp::croak("Can't use undef as constant name");
	}

	# Normal constant name
	if ($name =~ /^_?[^\W_0-9]\w*$str_end/ and !$forbidden{$name}) {
	    # Everything is okay

	# Name forced into main, but we're not in main. Fatal.
	} elsif ($forced_into_main{$name} and $pkg ne 'main') {
	    require Carp;
	    Carp::croak("Constant name '$name' is forced into main::");

	# Starts with double underscore. Fatal.
	} elsif ($name =~ /^__/) {
	    require Carp;
	    Carp::croak("Constant name '$name' begins with '__'");

	# Maybe the name is tolerable
	} elsif ($name =~ /^[A-Za-z_]\w*$str_end/) {
	    # Then we'll warn only if you've asked for warnings
	    if (warnings::enabled()) {
		if ($keywords{$name}) {
		    warnings::warn("Constant name '$name' is a Perl keyword");
		} elsif ($forced_into_main{$name}) {
		    warnings::warn("Constant name '$name' is " .
			"forced into package main::");
		}
	    }

	# Looks like a boolean
	# use constant FRED == fred;
	} elsif ($name =~ /^[01]?$str_end/) {
            require Carp;
	    if (@_) {
		Carp::croak("Constant name '$name' is invalid");
	    } else {
		Carp::croak("Constant name looks like boolean value");
	    }

	} else {
	   # Must have bad characters
            require Carp;
	    Carp::croak("Constant name '$name' has invalid characters");
	}

	{
	    no strict 'refs';
	    my $full_name = "${pkg}::$name";
	    $declared{$full_name}++;
	    if ($multiple || @_ == 1) {
		my $scalar = $multiple ? $constants->{$name} : $_[0];
		if ($symtab && !exists $symtab->{$name}) {
		    # No typeglob yet, so we can use a reference as space-
		    # efficient proxy for a constant subroutine
		    # The check in Perl_ck_rvconst knows that inlinable
		    # constants from cv_const_sv are read only. So we have to:
		    Internals::SvREADONLY($scalar, 1);
		    $symtab->{$name} = \$scalar;
		    mro::method_changed_in($pkg);
		} else {
		    *$full_name = sub () { $scalar };
		}
	    } elsif (@_) {
		my @list = @_;
		*$full_name = sub () { @list };
	    } else {
		*$full_name = sub () { };
	    }
	}
    }
}

1;

__END__

=head1 NAME

constant - Perl pragma to declare constants

=head1 SYNOPSIS

    use constant PI    => 4 * atan2(1, 1);
    use constant DEBUG => 0;

    print "Pi equals ", PI, "...\n" if DEBUG;

    use constant {
        SEC   => 0,
        MIN   => 1,
        HOUR  => 2,
        MDAY  => 3,
        MON   => 4,
        YEAR  => 5,
        WDAY  => 6,
        YDAY  => 7,
        ISDST => 8,
    };

    use constant WEEKDAYS => qw(
        Sunday Monday Tuesday Wednesday Thursday Friday Saturday
    );

    print "Today is ", (WEEKDAYS)[ (localtime)[WDAY] ], ".\n";

=head1 DESCRIPTION

This pragma allows you to declare constants at compile-time.

When you declare a constant such as C<PI> using the method shown
above, each machine your script runs upon can have as many digits
of accuracy as it can use. Also, your program will be easier to
read, more likely to be maintained (and maintained correctly), and
far less likely to send a space probe to the wrong planet because
nobody noticed the one equation in which you wrote C<3.14195>.

When a constant is used in an expression, perl replaces it with its
value at compile time, and may then optimize the expression further.
In particular, any code in an C<if (CONSTANT)> block will be optimized
away if the constant is false.

=head1 NOTES

As with all C<use> directives, defining a constant happens at
compile time. Thus, it's probably not correct to put a constant
declaration inside of a conditional statement (like C<if ($foo)
{ use constant ... }>).

Constants defined using this module cannot be interpolated into
strings like variables.  However, concatenation works just fine:

    print "Pi equals PI...\n";        # WRONG: does not expand "PI"
    print "Pi equals ".PI."...\n";    # right

Even though a reference may be declared as a constant, the reference may
point to data which may be changed, as this code shows.

    use constant ARRAY => [ 1,2,3,4 ];
    print ARRAY->[1];
    ARRAY->[1] = " be changed";
    print ARRAY->[1];

Dereferencing constant references incorrectly (such as using an array
subscript on a constant hash reference, or vice versa) will be trapped at
compile time.

Constants belong to the package they are defined in.  To refer to a
constant defined in another package, specify the full package name, as
in C<Some::Package::CONSTANT>.  Constants may be exported by modules,
and may also be called as either class or instance methods, that is,
as C<< Some::Package->CONSTANT >> or as C<< $obj->CONSTANT >> where
C<$obj> is an instance of C<Some::Package>.  Subclasses may define
their own constants to override those in their base class.

The use of all caps for constant names is merely a convention,
although it is recommended in order to make constants stand out
and to help avoid collisions with other barewords, keywords, and
subroutine names. Constant names must begin with a letter or
underscore. Names beginning with a double underscore are reserved. Some
poor choices for names will generate warnings, if warnings are enabled at
compile time.

=head2 List constants

Constants may be lists of more (or less) than one value.  A constant
with no values evaluates to C<undef> in scalar context.  Note that
constants with more than one value do I<not> return their last value in
scalar context as one might expect.  They currently return the number
of values, but B<this may change in the future>.  Do not use constants
with multiple values in scalar context.

B<NOTE:> This implies that the expression defining the value of a
constant is evaluated in list context.  This may produce surprises:

    use constant TIMESTAMP => localtime;                # WRONG!
    use constant TIMESTAMP => scalar localtime;         # right

The first line above defines C<TIMESTAMP> as a 9-element list, as
returned by C<localtime()> in list context.  To set it to the string
returned by C<localtime()> in scalar context, an explicit C<scalar>
keyword is required.

List constants are lists, not arrays.  To index or slice them, they
must be placed in parentheses.

    my @workdays = WEEKDAYS[1 .. 5];            # WRONG!
    my @workdays = (WEEKDAYS)[1 .. 5];          # right

=head2 Defining multiple constants at once

Instead of writing multiple C<use constant> statements, you may define
multiple constants in a single statement by giving, instead of the
constant name, a reference to a hash where the keys are the names of
the constants to be defined.  Obviously, all constants defined using
this method must have a single value.

    use constant {
        FOO => "A single value",
        BAR => "This", "won't", "work!",        # Error!
    };

This is a fundamental limitation of the way hashes are constructed in
Perl.  The error messages produced when this happens will often be
quite cryptic -- in the worst case there may be none at all, and
you'll only later find that something is broken.

When defining multiple constants, you cannot use the values of other
constants defined in the same declaration.  This is because the
calling package doesn't know about any constant within that group
until I<after> the C<use> statement is finished.

    use constant {
        BITMASK => 0xAFBAEBA8,
        NEGMASK => ~BITMASK,                    # Error!
    };

=head2 Magic constants

Magical values and references can be made into constants at compile
time, allowing for way cool stuff like this.  (These error numbers
aren't totally portable, alas.)

    use constant E2BIG => ($! = 7);
    print   E2BIG, "\n";        # something like "Arg list too long"
    print 0+E2BIG, "\n";        # "7"

You can't produce a tied constant by giving a tied scalar as the
value.  References to tied variables, however, can be used as
constants without any problems.

=head1 TECHNICAL NOTES

In the current implementation, scalar constants are actually
inlinable subroutines. As of version 5.004 of Perl, the appropriate
scalar constant is inserted directly in place of some subroutine
calls, thereby saving the overhead of a subroutine call. See
L<perlsub/"Constant Functions"> for details about how and when this
happens.

In the rare case in which you need to discover at run time whether a
particular constant has been declared via this module, you may use
this function to examine the hash C<%constant::declared>. If the given
constant name does not include a package name, the current package is
used.

    sub declared ($) {
        use constant 1.01;              # don't omit this!
        my $name = shift;
        $name =~ s/^::/main::/;
        my $pkg = caller;
        my $full_name = $name =~ /::/ ? $name : "${pkg}::$name";
        $constant::declared{$full_name};
    }

=head1 CAVEATS

In the current version of Perl, list constants are not inlined
and some symbols may be redefined without generating a warning.

It is not possible to have a subroutine or a keyword with the same
name as a constant in the same package. This is probably a Good Thing.

A constant with a name in the list C<STDIN STDOUT STDERR ARGV ARGVOUT
ENV INC SIG> is not allowed anywhere but in package C<main::>, for
technical reasons. 

Unlike constants in some languages, these cannot be overridden
on the command line or via environment variables.

You can get into trouble if you use constants in a context which
automatically quotes barewords (as is true for any subroutine call).
For example, you can't say C<$hash{CONSTANT}> because C<CONSTANT> will
be interpreted as a string.  Use C<$hash{CONSTANT()}> or
C<$hash{+CONSTANT}> to prevent the bareword quoting mechanism from
kicking in.  Similarly, since the C<< => >> operator quotes a bareword
immediately to its left, you have to say C<< CONSTANT() => 'value' >>
(or simply use a comma in place of the big arrow) instead of
C<< CONSTANT => 'value' >>.

=head1 BUGS

Please report any bugs or feature requests via the perlbug(1) utility.

=head1 AUTHORS

Tom Phoenix, E<lt>F<rootbeer@redcat.com>E<gt>, with help from
many other folks.

Multiple constant declarations at once added by Casey West,
E<lt>F<casey@geeknest.com>E<gt>.

Documentation mostly rewritten by Ilmari Karonen,
E<lt>F<perl@itz.pp.sci.fi>E<gt>.

This program is maintained by the Perl 5 Porters. 
The CPAN distribution is maintained by SE<eacute>bastien Aperghis-Tramoni
E<lt>F<sebastien@aperghis.net>E<gt>.

=head1 COPYRIGHT

Copyright (C) 1997, 1999 Tom Phoenix

This module is free software; you can redistribute it or modify it
under the same terms as Perl itself.

=cut
